Fix apush, aset and aunset for global php arrays

Chemaclass opened the pull request:
### 🤔 Background

Issue: https://github.com/phel-lang/phel-lang/issues/578

### 💡 Goal

Fix `php/push`, `php/aset` and `php/aunset` on native php arrays when using global definitions.

### 🔖 Changes

- Allowing using `getDefinitionReference()` that will be used when dealing with native php arrays

jenshaase commented:
The generate code is something like this:

```php
(\Phel\Lang\Registry::getInstance()->getDefinition("namespace", "arr"))[] = 4
```

The returned array for the `getDefinition` function is a copy of the array and not a reference.

Don't have a good solution yet.

Chemaclass commented:
Jap, realised that after creating this integration test (not comited [yet], just for debugging purposes):
```php
--PHEL--
(def arr (php/array 1 2 3))
(def new-arr (php/apush arr 4))
(is new-arr [1 2 3 4])
--PHP--
\Phel\Lang\Registry::getInstance()->addDefinition(
  "user",
  "arr",
  array(1, 2, 3),
  \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
    \Phel\Lang\Keyword::create("start-location"), \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
      \Phel\Lang\Keyword::create("file"), "Call/php-def-push.test",
      \Phel\Lang\Keyword::create("line"), 1,
      \Phel\Lang\Keyword::create("column"), 0
    ),
    \Phel\Lang\Keyword::create("end-location"), \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
      \Phel\Lang\Keyword::create("file"), "Call/php-def-push.test",
      \Phel\Lang\Keyword::create("line"), 1,
      \Phel\Lang\Keyword::create("column"), 27
    )
  )
);
\Phel\Lang\Registry::getInstance()->addDefinition(
  "user",
  "new-arr",
  (\Phel\Lang\Registry::getInstance()->getDefinition("user", "arr"))[] = 4,
  \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
    \Phel\Lang\Keyword::create("start-location"), \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
      \Phel\Lang\Keyword::create("file"), "Call/php-def-push.test",
      \Phel\Lang\Keyword::create("line"), 2,
      \Phel\Lang\Keyword::create("column"), 0
    ),
    \Phel\Lang\Keyword::create("end-location"), \Phel\Lang\TypeFactory::getInstance()->persistentMapFromKVs(
      \Phel\Lang\Keyword::create("file"), "Call/php-def-push.test",
      \Phel\Lang\Keyword::create("line"), 2,
      \Phel\Lang\Keyword::create("column"), 31
    )
  )
);
\Phel\Lang\Registry::getInstance()->getDefinition("user", "new-arr");
```

Would it be possible to define in the `PhpArrayPushEmitter` (or maybe in an earlier stage) that we mutate the copy array and then we set the copy array into the old one (overriding the previous reference)?

jenshaase commented:
`addDefinition` should not update anything. Definition should only be set once.

However, there is a workaround using `var`. This is even better Phel code because it explicitly says that there is some mutation happening:

```
phel:1> (def arr (var (php/array 1 2 3)))
1
phel:2> (deref arr)
<PHP-Array [1, 2, 3]>
phel:3> (let [a (deref arr)] (php/apush a 4) (set! arr a))
nil
phel:4> (deref arr)
<PHP-Array [1, 2, 3, 4]>
phel:5> 
```

Chemaclass commented:
What do you think about this ☝🏼 about using `(var)`, @cgrabenstein (related to https://github.com/phel-lang/phel-lang/issues/578)

jenshaase commented:
Is there a way to have a `getDefinitionReference` method that returns the reference and not the copy?

Chemaclass commented:
@jenshaase there might be a way with https://www.php.net/manual/en/language.references.return.php 🤔

cgrabenstein commented:
@Chemaclass using `var` is a reasonable workaround. However _if_ there's an implementation that allows to use `apush` without `var`, I would prefer that. I see that this goes against the way how php handles arrays and will require some shenanigans though. 

@jenshaase it's totally possible. See https://3v4l.org/sYHa8

jenshaase commented:
I added a fix. @Chemaclass do you want to review it?

jenshaase commented:
I also fixed it for `php/aset` and `php/aunset`

Chemaclass commented:
Thanks @jenshaase for your help! I added some unit tests for the Registry logic, and renamed some variables. 
Can you review, approve and merge the PR? I cannot approve it because I am the author of the PR 😝
